### PR TITLE
Promotions: Deprecate Spree::PromotionRule#actionable and split rules that implement it

### PR DESCRIPTION
**Pull request by mamhoff**
The API of promotion rules is a bit strange: There's the `applicable` method, which looks like promotion rules could apply to line items as well as orders, and there's `actionable?`, which is like `eligible?`, but for line items. 

The `Product` rule is really two rules in one: An order rule (Does this order contain product x?) and a line item rule (Is this line item product x?). 

Same with the `Taxon` rule: 
- Does the order contain a product with taxon x?
- Does the line item have a product with taxon x?

Same with the `OptionValue` rule:
- Does the order contain a variant with option value X?
- Does the line item have a variant with option value X?

What this PR does is 
- split those rules up into two
- migrate people's promotions that use these rules such that they now have the additional "line item" rule as well, preserving behavior in existing stores
- deprecate `Spree::Promotion#line_item_actionable?`
- deprecate `Spree::PromotionRule#actionable?` 

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [x] I've changed the guides to reflect my changes

### CODE REVIEW

**Review by jarednorman — APPROVED** [review]
This looks good, though we could additionally deprecate the old rule if we want, right?

**Comment by mamhoff** [comment]
I've found a bug in the promotion system that makes line-item level rules do things we don't want them to do, so I'm marking this as  draft until I've got the fix for that into a state that works. I'm likely also going to split the other two rules that have `actionable?` and deprecate that and related methods. There is a little bit of UI work left that should help admins understand the new system a lot better, as well. 

In this PR, I'm moving the `Product` rule to an `OrderProduct` rule, but given that this is an STI class, and deprecating classes requires changes to the `type` column, I think I'm rather going to build an upgrade path that simply adds the line items rules to any promotions that have rules with `actionable?`.

**Comment by mamhoff** [comment]
Hm, this got a bit bigger, but I think I got to a good place with this. I'm not deprecating the old rules; instead I leave them to do what their names and documentation and implementation say that they do: Check order eligibility. I added new rules for line item eligibility, and deprecated the use of `actionable?` for line items. 

This change gives meaning to the `applicable?` method, and it allows to extend the promotion system to also support checks for e.g. shipments or shipping rates in the future.

**Comment by mamhoff** [comment]
@jarednorman I've re-requested your review because this now contains a bunch more change than what you initially reviewed.

**Review by tvdeyen — COMMENTED** [review]
Very nice work. Thank you for removing some of the confusing parts of the promotion system.

File: core/app/models/spree/promotion/rules/line_item_product.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+module Spree
+  class Promotion < Spree::Base
+    module Rules
+      # A rule to apply a promotion only to line items with or without a chosen product
+      class LineItemProduct < Spree::PromotionRule
+        MATCH_POLICIES = %w(normal inverse)
```
[the remark this answers is not in the document]
Comment (mamhoff) [reply core/app/models/spree/promotion/rules/line_item_product.rb]: ok

File: core/app/models/spree/promotion/rules/option_value.rb [hunk 2]
```diff
@@ -11,14 +11,9 @@ def applicable?(promotable)
         end
 
         def eligible?(promotable, _options = {})
```
[the remark this answers is not in the document]
Comment (mamhoff) [reply core/app/models/spree/promotion/rules/option_value.rb]: done

File: core/config/locales/en.yml [hunk 3]
```diff
@@ -658,7 +658,7 @@ en:
       spree/promotion/rules/nth_order: Nth Order
       spree/promotion/rules/one_use_per_user: One Use Per User
       spree/promotion/rules/option_value: Option Value(s)
-      spree/promotion/rules/product: Product(s)
+      spree/promotion/rules/product: Order Product(s)
```
[the remark this answers is not in the document]
Comment (mamhoff) [reply core/config/locales/en.yml:667]: fixed

**Comment by mamhoff** [comment]
@kennyadsl would you mind having a look? Would this also be a change for 4.0?

**Comment by kennyadsl** [comment]
@mamhoff nope, if the public methods are deprecated and there's a clear transition path for who needs to change something in their codebase, I think we can marge this before 4.0 and only remove the deprecated parts in that major. 

Still, I didn't have time to review this PR yet, it's a bit large and requires some effort. Do you think there's a way or makes sense to split this into smaller PRs.

**Comment by mamhoff** [comment]
@kennyadsl I don't think I can meaningfully split it into smaller pull requests. I'm depecrating an API here, and I'm providing a replacement. I'm also implementing this change across the promotion rules we have. This will necessarily take a few LOC, but splitting it up over multiple PRs I think would hurt readability and understandability of what's going on. Sorry! But do take your time. Best!

**Comment by kennyadsl** [comment]
Update: I had a quick look and this looks very good. I also asked some team members to check if this would impact the stores they are working on. I will approve once I have positive answers. In the meantime, thanks for the fantastic work, Martin!

**Comment by mamhoff** [comment]
Closing in favor of https://github.com/friendlycart/solidus_friendly_promotions/

**Pull request closed without merging**